### ISSUE TITLE
ListView shifts down when another view appears

### ISSUE THREAD

**Issue by alexzaitsev** [bumptech/glide#1654]
Hello!

```
Glide version: 3.7.0
no integrations
```

I have an adapter. It's inherited from BaseAdapter. It has a lot of view types and one of them is gifs downloaded with Glide. The issue is that if the item with such image is visible - the list will be shifted when another view appears on the top of the list (in my case it's an advertisement). Let me illustrate my words with a couple of videos.

[First video](https://youtu.be/m2_8D6xAfQc) consist of 2 parts: at first I've opened a chat where we can observe the bug, at second I've opened another chat to show you that we have no bug there. Bug only occurs if a chat contain visible gifs. 
[Second video](https://youtu.be/3R1KEYpEoCY) demonstrates the same chat but all the lines of the function `loadGif` are commented so Glide isn't working here. And voila - no shifting. 
Excuse me for the ugly quality I can re-download the videos if it's needed.

My adapter is huge, here are some parts of it:
```
    private View.OnLayoutChangeListener onLayoutChangeListener = new View.OnLayoutChangeListener() {
        @Override
        public void onLayoutChange(View view, int left, int top, int right, int bottom,
                                   int oldLeft, int oldTop, int oldRight, int oldBottom) {
            ViewData viewData = (ViewData) view.getTag();
            if (viewData != null &&
                    (viewData.lastWidth != view.getWidth() ||
                            viewData.lastHeight != view.getHeight())) {
                // if view size was changed - reload gif
                loadGif();
                // save latest changes
                viewData.lastWidth = view.getWidth();
                viewData.lastHeight = view.getHeight();
                view.setTag(viewData);
            }
        }
    };

    private void startLoading() {
        if (currentGif == null || TextUtils.isEmpty(currentGif.getUrl()) ||
                currentGif.getWidth() == 0 || currentGif.getHeight() == 0) {
            onGifError();
            return;
        }

        setRetryVisible(false);
        int viewHeight = currentGif.getHeight() * loadedSizePx / currentGif.getWidth();

        ViewData tag = new ViewData(currentGif.getUrl(), currentGif.getWidth(), currentGif.getHeight());
        holder.setTag(tag);

        holder.addOnLayoutChangeListener(onLayoutChangeListener);

        if (holder.getWidth() == loadedSizePx && holder.getHeight() == viewHeight) {
            // view already has appropriate sizes
            loadGif();
        } else {
            setViewSize(holder, loadedSizePx, viewHeight + context.getResources().getDimensionPixelSize(R.dimen.gif_time_height));
        }
    }

    private void setViewSize(View view, int width, int height) {
        RelativeLayout.LayoutParams lp = (RelativeLayout.LayoutParams) view.getLayoutParams();
        lp.width = width;
        lp.height = height;
        view.setLayoutParams(lp);
    }

    private void loadGif() {
        ImageUtils.downloadGif(context, currentGifLoadType, currentGif.getUrl(), target,
                ImageUtils.GifScaleType.CENTER_CROP,
                new RequestListener<String, GlideDrawable>() {
                    @Override
                    public boolean onException(Exception e, String model, Target<GlideDrawable> target, boolean isFirstResource) {
                        proceedLoadQueue();
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(GlideDrawable resource, String model, Target<GlideDrawable> t,
                                                   boolean isFromMemoryCache, boolean isFirstResource) {
                        onGifSuccess();
                        return false;
                    }
                });
    }

    private void onGifError() {
        cacheGifState(false);
        progress.setVisibility(View.GONE);
        holder.setTag(null);
        target.setOnClickListener(null);
        target.setImageResource(context.getResources().getDisplayMetrics().density <= 1 ?
                R.drawable.sticker_unknown_mdpi : R.drawable.sticker_unknown);
        unknownText.setVisibility(View.VISIBLE);
    }

    private void onGifSuccess() {
        cacheGifState(true);
        progress.setVisibility(View.GONE);
        target.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(context, GifActivity.class);
                intent.putExtras(((NativeChatActivity) context).getIntent().getExtras());
                intent.putExtra(GifListActivity.EXTRA_GIF, gif);
                intent.putExtra(GifActivity.EXTRA_PREVIEW, true);
                context.startActivity(intent);
            }
        });
        unknownText.setVisibility(View.GONE);
    }
```

And ImageUtils:
```
public class ImageUtils {

    public enum GifScaleType {
        CENTER_CROP,
        FIT_CENTER
    }

    public static void downloadGif(Context context, GifLoadType gifLoadType, String url,
                                   ImageView target, GifScaleType scaleType,
                                   RequestListener<String, GlideDrawable> listenerGif) {
        downloadGif(context, gifLoadType, url, target, scaleType, listenerGif, null);
    }

    public static void downloadGif(Context context, GifLoadType gifLoadType, String url,
                                   ImageView target, GifScaleType scaleType,
                                   RequestListener<String, GlideDrawable> listenerGif,
                                   Priority priority) {
        GlideDrawableImageViewTarget imageViewTarget = new GlideDrawableImageViewTarget(target);
        DrawableRequestBuilder<String> gifRequestBuilder;
        if (gifLoadType.isCache()) {
            gifRequestBuilder = Glide.with(context).using(new NetworkDisablingLoader()).load(url);
        } else {
            gifRequestBuilder = Glide.with(context).load(url);
        }
        gifRequestBuilder
                .crossFade()
                .diskCacheStrategy(DiskCacheStrategy.SOURCE);
        if (listenerGif != null) {
            gifRequestBuilder.listener(listenerGif);
        }
        if (scaleType != null) {
            switch (scaleType) {
                case FIT_CENTER:
                    gifRequestBuilder.fitCenter();
                    break;
                case CENTER_CROP:
                    gifRequestBuilder.centerCrop();
                    break;
            }
        }
        gifRequestBuilder.priority(priority == null ? Priority.LOW : priority);
        gifRequestBuilder.into(imageViewTarget);
    }

    private static class NetworkDisablingLoader implements StreamModelLoader<String> {
        @Override
        public DataFetcher<InputStream> getResourceFetcher(final String model, int width, int height) {
            return new DataFetcher<InputStream>() {
                @Override
                public InputStream loadData(Priority priority) throws Exception {
                    throw new IOException("Forced Glide network failure");
                }

                @Override
                public void cleanup() {
                }

                @Override
                public String getId() {
                    return model;
                }

                @Override
                public void cancel() {
                }
            };
        }
    }
}
```
Finally here is my layout. It's big and probably not the best but I'm attaching it because you may need it: [gist](https://gist.github.com/alexzaitsev/5b498b64ed7ab69dc833a47afc37d8f7).

Please push me in the right direction, I've spent several days trying to fix this and need your help.

**Comment by alexzaitsev** [comment]
Hello, could you help me? I don't understand why it's happening only after Glide loads gif to the ImageView.

**Comment by alexzaitsev** [comment]
Hello, can you give me any explanation of the issue? I cannot release without it. Thank you.

**Comment by foqq** [comment]
you could try commented out  progress.setVisibility(View.GONE); and unknownText.setVisibility(View.GONE);

**Comment by alexzaitsev** [comment]
tried Fresco and faced with the same issue

**Comment by TWiStErRob** [comment]
@alexzaitsev sorry for not replying. I just took a look and it's really complex. Would need a repro app to investigate more. I suggest you swap `loadGif()` and updating the size to make sure there's no double-load going on.

I noticed two things in the first video:
 * GIFs are reloaded with a (probably) different size when the ad shows up; notice how they show progress bar (YT allows 0.25x speed playback).
 * The scrollbar gets smaller when the ad is shown, which suggest that it's part of the list (headerView?). Maybe try moving it outside the list?
